daft/filesystem.py

```diff
@@ -353,3 +369,86 @@ def join_path(fs: pafs.FileSystem, base_path: str, *sub_paths: str) -> str:
         return os.path.join(base_path, *sub_paths)
     else:
         return f"{base_path.rstrip('/')}/{'/'.join(sub_paths)}"
+
+
+###
+# File Overwriting
+###
+
+
+def _infer_fssspec_filesystem_from_protocol(
+    protocol: str,
+    io_config: IOConfig | None,
+) -> fsspec.AbstractFileSystem:
+    """
+    Infer the fsspec filesystem based on the protocol.
+    """
+
+    def _set_if_not_none(kwargs: dict[str, Any], key: str, val: Any | None):
+        """Helper method used when setting kwargs for fsspec filesystems"""
+        if val is not None:
+            kwargs[key] = val
+
+    kwargs: dict[str, Any] = {}
+    if protocol == "s3":
+        if io_config is not None and io_config.s3 is not None:
+            s3_config = io_config.s3
+            _set_if_not_none(kwargs, "key", s3_config.key_id)
+            _set_if_not_none(kwargs, "secret", s3_config.access_key)
+            _set_if_not_none(kwargs, "token", s3_config.session_token)
+            _set_if_not_none(kwargs, "client_kwargs", {"region_name": s3_config.region_name})
+            _set_if_not_none(kwargs, "endpoint_url", s3_config.endpoint_url)
+            _set_if_not_none(kwargs, "anon", s3_config.anonymous)
+        fs = fsspec.filesystem("s3", **kwargs)
+
+    elif protocol == "file":
+        fs = fsspec.filesystem("file")
+    elif protocol in {"gs", "gcs"}:
+        if io_config is not None and io_config.gcs is not None:
+            gcs_config = io_config.gcs
+            _set_if_not_none(kwargs, "token", gcs_config.token)
+            _set_if_not_none(kwargs, "project", gcs_config.project_id)
+        fs = fsspec.filesystem("gs", **kwargs)
+    elif protocol in {"az", "abfs", "abfss"}:
+        if io_config is not None and io_config.azure is not None:
+            azure_config = io_config.azure
+            _set_if_not_none(kwargs, "account_name", azure_config.storage_account)
+            _set_if_not_none(kwargs, "account_key", azure_config.access_key)
+            _set_if_not_none(kwargs, "sas_token", azure_config.sas_token)
+            _set_if_not_none(kwargs, "tenant_id", azure_config.tenant_id)
+            _set_if_not_none(kwargs, "client_id", azure_config.client_id)
+            _set_if_not_none(kwargs, "client_secret", azure_config.client_secret)
+            _set_if_not_none(kwargs, "anon", azure_config.anonymous)
+        fs = fsspec.filesystem("abfs", **kwargs)
+    else:
+        raise NotImplementedError(f"Cannot infer Fsspec filesystem for protocol {protocol}: please file an issue!")
+
+    return fs
```

Comment:
I would avoid fsspec if possible here so we can avoid taking a dependency on it.

I'm also not sure if the bulk delete API has any performance benefit over a serial delete call... I guess for the services (e.g. S3) we can parallelize `DELETE` requests over the wire.

Ideally we can support `DELETE` on our own IO clients, but in the absence of that shall we use PyArrow instead and just naively delete one-by-one?

Reply:
Yeah we can use PyArrow, should be easier to implement anyways since we already have the machinery to infer a pyarrow filesystem. I can also run some tests to see if paralellizing the deletes make sense.